src/vos/vos_obj_flat.h

```diff
@@ -0,0 +1,128 @@
+/**
+ * (C) Copyright 2023-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * Object flatten related API and structures
+ * src/vos/vos_obj_flat.h
+ */
+
+#ifndef __VOS_OBJ_FLAT_H__
+#define __VOS_OBJ_FLAT_H__
+
+#include "vos_layout.h"
+
+enum {
+	VOF_DKEY	= 0x1,
+	VOF_AKEY	= 0x2,
+	VOF_SINGV	= 0x3,
+	VOF_EXT		= 0x4,
+};
+
+#define VOF_KEY_INLINE_SZ	(4)
+#define VOF_SINGV_INLINE_SZ	(8)
+
+/** dkey/akey/value for VOS flattened object, 20B each */
+struct vof_item_df {
+	/** type (dkey/akey/sing-value/array-ext) */
+	uint16_t				ofk_type;
+	/** number of checksums */
+	uint16_t				ofk_cs_nr;
+	/** size of the item, for single-value is the local record size */
+	uint32_t				ofk_size;
+	union {
+		/* dkey/akey */
+		struct {
+			union {
+				/** offset of the key */
+				uint32_t	ofk_key_off;
+				/** inline key (for short key and no key csum) */
+				uint8_t		ofk_key[VOF_KEY_INLINE_SZ];
+			};
+			/** offset of children (akey or value) */
+			uint32_t		ofk_child_off;
+			/** number of children */
+			uint16_t		ofk_child_nr;
+			/** type of children */
+			uint16_t		ofk_child_type;
+		};
+		/* array extent */
+		struct {
+			/** index of the value extent */
+			uint64_t		ofk_ext_idx;
+			/** offset of value */
+			uint32_t		ofk_ext_off;
+		};
+		/* single-value */
+		struct {
+			/* global record size of the single-value */
+			uint32_t		ofk_singv_gsize;
+			union {
+				/** offset of the single-value */
+				uint32_t		ofk_singv_off;
+				/** inline data (for short single-value and no csum) */
+				uint8_t			ofk_singv[VOF_SINGV_INLINE_SZ];
+			};
+		};
+	};
+} __attribute__ ((aligned(4)));
```

Comment:
does this mean we always align to 4 bytes even if it's a 64-bit value in middle of struct?  Do we ever plan to access these directly?

Reply:
I was try to put it to 20B but just realized that the size of record should be stored for each array ext, so will change it to 24B, so can avoid that problem.